src/main/kotlin/org/jitsi/videobridge/transport/ice/IceTransport.kt

```diff
@@ -0,0 +1,443 @@
+package org.jitsi.videobridge.transport.ice
+
+import org.ice4j.Transport
+import org.ice4j.TransportAddress
+import org.ice4j.ice.Agent
+import org.ice4j.ice.CandidateType
+import org.ice4j.ice.IceMediaStream
+import org.ice4j.ice.IceProcessingState
+import org.ice4j.ice.LocalCandidate
+import org.ice4j.ice.RemoteCandidate
+import org.ice4j.socket.SocketClosedException
+import org.jitsi.nlj.util.OrderedJsonObject
+import org.jitsi.utils.logging2.Logger
+import org.jitsi.utils.logging2.cdebug
+import org.jitsi.utils.logging2.createChildLogger
+import org.jitsi.videobridge.ice.Harvesters
+import org.jitsi.videobridge.ice.IceConfig.Config
+import org.jitsi.videobridge.ice.TransportUtils
+import org.jitsi.videobridge.util.ByteBufferPool
+import org.jitsi.xmpp.extensions.jingle.CandidatePacketExtension
+import org.jitsi.xmpp.extensions.jingle.IceUdpTransportPacketExtension
+import org.jitsi.xmpp.extensions.jingle.RtcpmuxPacketExtension
+import sun.net.util.IPAddressUtil
+import java.beans.PropertyChangeEvent
+import java.io.IOException
+import java.net.DatagramPacket
+import java.time.Clock
+import java.time.Instant
+import java.util.concurrent.atomic.AtomicBoolean
+
+class IceTransport @JvmOverloads constructor(
+    id: String,
+    /**
+     * Whether or not the ICE agent created by this transport should be the
+     * 'controlling' role.
+     */
+    controlling: Boolean,
+    parentLogger: Logger,
+    private val clock: Clock = Clock.systemUTC()
+) {
+    private val logger = createChildLogger(parentLogger)
+
+    /**
+     * The handler which will be invoked when data is received.  The handler
+     * does *not* own the buffer passed to it, so a copy must be made if it wants
+     * to use the data after the handler call finishes.  This field should be
+     * set by some other entity which wishes to handle the incoming data
+     * received over the ICE connection.
+     * NOTE: we don't create a packet in [IceTransport] because
+     * RTP packets want space before and after and [IceTransport]
+     * has no notion of what kind of data is contained within the buffer.
+     */
+    @JvmField
+    var incomingDataHandler: IncomingDataHandler? = null
+
+    /**
+     * The handler which will be invoked when events fired by [IceTransport]
+     * occur.  This field should be set by another entity who wishes to handle
+     * the events.  Handlers will only be notified of events which occur
+     * *after* the handler has been set.
+     */
+    @JvmField
+    var eventHandler: EventHandler? = null
+
+    /**
+     * Whether or not this [IceTransport] has connected.
+     */
+    private val iceConnected = AtomicBoolean(false)
+    /**
+     * Whether or not this [IceTransport] has failed to connect.
+     */
+    private val iceFailed = AtomicBoolean(false)
+
+    fun hasFailed(): Boolean = iceFailed.get()
+
+    fun isConnected(): Boolean = iceConnected.get()

```

Comment:
Having `hasFailed` and `isConnected` as separate booleans seems odd -- e.g. there's no way they can both be true at the same time, can there?  Should this be some sort of enum instead?

Reply:
They can't both be true, but they can both be false :)

Reply:
Yeah, that's why I was thinking some kind of enum state might work better.

Possibly the same states as `IceProcessingState`, though I don't know if this needs the `COMPLETED`/`TERMINATED` distinction.

Reply:
So something like `RUNNING`, `FAILED`, `SUCCEEDED` ?

Reply:
Yeah.  (Is there ever a case where this will be in a `WAITING` state, and if so does this API care?)

Reply:
ICE restarts are tricky, because usually you want the `RUNNING` -> `SUCCEEDED` transition to gate things like starting the DTLS handshake, and you don't have to do that for ICE restarts.

Reply:
I'm good with the enum change, but I was trying to change as little other code as possible as part of this PR (and the ensuing PRs) because they include such fundamental code I didn't want to make the PR bigger/touch anything I didn't have to.

Reply:
> ICE restarts are tricky, because usually you want the `RUNNING` -> `SUCCEEDED` transition to gate things like starting the DTLS handshake, and you don't have to do that for ICE restarts.

Yeah, and I think if we _did_ need to trigger something, we can use the events from `IceTransport` (since we'll fire a new `connected` event in that case)

Reply:
Ok, yeah, if this is just the Kotlin-ization of the existing code it's fine. (Because of the Kotlin translation it's hard to tell how much is new in `IceTransport`.)